### DISCUSSION TITLE
[solved] Stuck on Scala 2.13.3 -> How to upgrade?

### DISCUSSION THREAD

**Discussion by lefou** [com-lihaoyi/mill#1278] [Development]
Hi all, we are currently stuck at Scala 2.13.3, as for any bump to 2.13.4 or 2.13.5 our CI fails with a `ConcurrentModificationException` thrown by zinc. I, for some reasons, didn't even figured out how to reproduce it locally and also wasn't able to communicate our/my problem in the sbt channel properly.

Here are some pointers: com-lihaoyi/mill#1116, com-lihaoyi/mill#1185, com-lihaoyi/mill#1197, scala/scala#9174

Any help would be really appreciated. :pray: 

_First posted on Gitter: https://gitter.im/lihaoyi/mill?at=60769326c60826673ba24516_

**Comment by lefou** [comment]
TweKei made some progress in isolating the issue (posted on Gitter):

> TzeKei Lee @chikei 08:38
> maybe we can refine further (like not using ammonite or something may change classloader ) but this gist https://gist.github.com/chikei/413956ffbf737402a3ae96f2010635c7 reproduce it under ammonite repl 2.3.8-47-7e8c5dd3 with scala 2.13 and with empty out/main/client/compile/dest folder.

Seems to be related to our Java code in mill client.

**Comment by chikei** [comment]
Here is the minimal example under sbt https://github.com/chikei/zinc-ConcurrentModificationException , it looks like an actual zinc bug. Interestingly I've tried extends some other jdk class and only `Thread` cause the problem, maybe we can just rewrite `FileToStreamTailer` without directly extends thread to work around.

**Reply by lefou** [reply]
Great, I can now reproduce the issue. It's also a nice small example so that a zinc developer can review if we use the zinc compiler API correctly. I'll play a bit with it to see, if we can work around the issue. And I guess, we should report this upstream.

**Reply by lefou** [reply]
I opened an Issue in sbt/zinc: https://github.com/sbt/zinc/issues/974

**Reply by lefou** [reply]
My PR https://github.com/sbt/zinc/pull/975 got merged. I hope a new zinc release is cut soon, so that we can update the Scala versions in mill too.

**Reply by lefou** [reply]
See #1197

**Reply by lefou** [reply]
Meanwhile, we have mill 0.9.8 released, which is build on Scala 2.13.5!

@chikei Thank you very much for your help! Without it, we couldn't report and identify the issue in zinc that easy. :pray:
